python-frontend/src/main/java/org/sonar/python/semantic/v2/TypeInferenceV2.java

```diff
@@ -85,6 +91,35 @@ public void visitStringLiteral(StringLiteral stringLiteral) {
     ((StringLiteralImpl) stringLiteral).typeV2(new ObjectType(strType, List.of(), List.of()));
   }
 
+  @Override
+  public void visitTuple(Tuple tuple) {
+    super.visitTuple(tuple);
+    List<PythonType> contentTypes = tuple.elements().stream().map(Expression::typeV2).distinct().toList();
+    List<PythonType> attributes = List.of();
+    if (contentTypes.size() == 1 && !contentTypes.get(0).equals(PythonType.UNKNOWN)) {
+      attributes = contentTypes;
+    }
+    ModuleType builtins = this.projectLevelTypeTable.getModule(BUILTINS);
+    PythonType tupleType = builtins.resolveMember("tuple").orElse(PythonType.UNKNOWN);
+    ((TupleImpl) tuple).typeV2(new ObjectType(tupleType,  attributes, List.of()));
```

Comment:
using `List.of()` could be dangerous since it is immutable. I'd prefer to use array lists instead